### PR TITLE
Fixes #148

### LINKED ISSUE #148: critwall: not running in correct security context with Subject.doAs

### PR DESCRIPTION
**Pull request by at88mph**
Also ensure that exceptions are passed up.

### CODE REVIEW

File: critwall/src/main/java/org/opencadc/critwall/FileSyncJob.java [hunk 1]
```diff
@@ -126,16 +131,32 @@ public FileSyncJob(URI artifactID, URI locatorServiceID, StorageAdapter storageA
...
+
     @Override
     public void run() {
         log.info("fileSyncJob start.");
+        if (this.owner == null) {
+            doSync();
+        } else {
+            Subject.doAs(this.owner, (PrivilegedAction<Void>) () -> {
+                doSync();
+                return null;
```
Comment (pdowler) [inline critwall/src/main/java/org/opencadc/critwall/FileSyncJob.java:146]: Should not have a `return null` here. Surprised compiler didn't complain...

File: critwall/src/main/java/org/opencadc/critwall/FileSyncJob.java [hunk 2]
```diff
@@ -126,16 +131,32 @@ public FileSyncJob(URI artifactID, URI locatorServiceID, StorageAdapter storageA
...
+        log.info("doSync start.");
         final List<URL> urlList;
         try {
-            urlList = getdownloadURLs(this.locatorService, this.artifactID);
+            urlList = getDownloadURLs(this.locatorService, this.artifactID);
         } catch (Exception e) {
             // fail - nothing can be done without the list, negotiation cannot be retried
             log.error("transfer negotiation failed.");
-            return;
+            throw new RuntimeException(e);
```
Comment (pdowler) [inline critwall/src/main/java/org/opencadc/critwall/FileSyncJob.java:159]: Can you please put the causing exception in the log.error call? When this failure happens all detailed info is lost.

File: critwall/src/main/java/org/opencadc/critwall/Main.java [hunk 3]
```diff
@@ -229,8 +237,31 @@ public static void main(String[] args) {
                 "org.postgresql.Driver",
                 dbUrl);
 
-            FileSync doit = new FileSync(daoConfig, cc, localStorage, locatorService, bucketSel, nthreads);
-            doit.run();
+            final Subject subject = SSLUtil.createSubject(new File(CERTIFICATE_FILE_LOCATION));
```
Comment (pdowler) [inline critwall/src/main/java/org/opencadc/critwall/Main.java:240]: Two things here:
1. The application should work without a certificate so this needs to check if the file exists before creating a Subject.
2. The FileSync ctor should not do any remote calls that require auth, so only the `run()` needs to be inside the Subject.doAs. It's bad practice to have setup (ctors) require auth so we want to find and fix that kind of thing.
3. There is a standard wrapper (RunnableAction) you can use to execute a Runnable inside a Subject.doAs

**Review by pdowler — CHANGES_REQUESTED** [review]

File: critwall/src/main/java/org/opencadc/critwall/FileSyncJob.java (hunk 1 above)
Comment (at88mph) [reply critwall/src/main/java/org/opencadc/critwall/FileSyncJob.java:146]: The `PrivilegedAction` needs to return something though.  Is there a different way to do it?

Comment (at88mph) [reply critwall/src/main/java/org/opencadc/critwall/FileSyncJob.java:146]: Actually, changed to be a `RunnableAction` instead.

**Review by pdowler — APPROVED** [review]

**Pull request merged**